stan-dev/math
# Bugfix/issue 487 hypergeometric functions

#### Submisison Checklist


#### Summary:

There are currently bugs in F32 and ... I think it's grad_F32, and some potential bugs in grad_2F1 that I haven't demonstrated yet (it _can_ terminate at an arbitrary point).  See issue #487.  

#### Intended Effect:

Fix bug in F32 and grad_F32, add some doc and tests.

#### How to Verify:

Run new tests versus `develop` to see them fail, they pass with this branch.  On `develop` at least one of the tests hangs on my machine.  Test values come from Mathematica and R's hypergeom package.  The tests that fail against develop are:

```
cd ${STAN-MATH}
git checkout develop
git pull origin develop
git checkout bugfix/issue-487-hypergeometric-functions -- test/unit/math/prim/scal/fun/F32_test.cpp
make test/unit/math/prim/scal/fun/F32_test
./test/unit/math/prim/scal/fun/F32_test
git checkout bugfix/issue-487-hypergeometric-functions -- test/unit/math/prim/scal/prob/beta_binomial_cdf_log_test.cpp
make tests/unit/math/prim/scal/prob/beta_binomial_cdf_log_test
./test/unit/math/prim/scal/prob/beta_binomial_cdf_log_test
```

#### Side Effects:

betabinomial_cdf is now more correct and has a test that should fail with the code on `develop` but pass with this branch.  The test seems to represent reasonable parameter values. 

#### Documentation:

Added where it was missing.

#### Reviewer Suggestions: 

- The tests for F32 are done as far as I'm concerned and I was going to add parallel ones for grad_F32 and grad_2F1.  If you want anymore let me know
- The logic tracking minus signs seems bulky, suggestions appreciated but it is as hairy as it looks 
- F32 throws when the accumulator is infinite because that's a known non-convergence that sometimes occurs long before 

#### Copyright and Licensing

Copyright holder: Krzysztof Sakrejda

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
